pool.go

```diff
@@ -161,26 +180,69 @@ func (p *pool) fetchWith(ctx context.Context, c cid.Cid, with string) (blk block
 	<-p.started
 
 	left := p.config.MaxRetries
-	if left == 0 {
-		left = DefaultMaxRetries
+
+	aff := with
+	if aff == "" {
+		aff = c.Hash().B58String()
 	}
 
-	for i := 0; i < left; i++ {
-		aff := with
-		if aff == "" {
-			aff = fmt.Sprintf("%d%s", i, c.Hash().B58String())
-		} else {
-			aff = fmt.Sprintf("%d%s", i, aff)
-		}
-		p.lk.RLock()
-		member := p.c.LocateKey([]byte(aff))
+	p.lk.RLock()
+	if left > len(p.endpoints) {
+		left = len(p.endpoints)
+	}
+	if p.c == nil || p.c.Size() == 0 {
 		p.lk.RUnlock()
-		root := member.String()
+		return nil, ErrNoBackend
+	}
+	nodes, ok := p.c.GetNodes(aff, left)
+	p.lk.RUnlock()
+	if !ok || len(nodes) == 0 {
+		select {
+		case p.refresh <- struct{}{}:
+		default:
+		}
+
+		return nil, ErrNoBackend
+	}
 
-		blk, err = p.doFetch(ctx, root, c)
+	for i := 0; i < len(nodes); i++ {
+		blk, err = p.doFetch(ctx, nodes[i], c)
```

Comment:
If you have no timeout for `doFetch` here, you effectively have no retries.

In my tests, either saturn PoP replies fast with block, or it hangs until [60s timeout in biforst-gateway](https://github.com/ipfs/bifrost-gateway/pull/33) hits, and cancels the entire context. There is never a scenario when `MaxRetries` logic gets executed here.

Some ideas: 
- if you introduce  timeout per `doFetch` attempt, maybe 19s (so 3 retries take  <60s) (we already downvote on each failure)
- try at least 2 PoPs in parallel, and use the fastest response, cancelling remaining ones (+ boost weight of the fast one)

Reply:
Addressed in https://github.com/filecoin-saturn/caboose/pull/26/commits/c13c0132bf17f1fd79295c5d15723725eca5d941.